### PR TITLE
Ability to update registered solver in LivepeerVerifier

### PR DESCRIPTION
**Pull request by yondonfu**
This LIP proposes an upgrade to the LivepeerVerifier contract that allows for both the addition and removal of solver addresses. The proposal also describes some related state management updates that can simplify the logic of the contract.

### CODE REVIEW

File: LIPs/LIP-update-solver.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
... to defend against the following timing attack: if a malicious transcoder knew `upgradeBlock`, it could skip performing any real transcoding work, submit a fraudulent transcode claim with transcode receipts containing dummy data and submit challenged segments for verification right before `upgradeBlock` knowing that it will fail verification on the challenged segments. Since none of the state from `LivepeerVerifierV1` is migrated to `LivepeerVerifierV2`, the newly upgraded contract will not have a record of the verification requests created by the transcoder when it submitted challenged segments for verification right before `upgradeBlock`. As a result, the registered solver will not be able to submit computation results to prove the malicious transcoder failed verification on the challenged segments, thus allowing the malicious transcoder to extract fees from a broadcaster without performing real transcoding work. If transcoders do not know the value of `upgradeBlock`, this timing attack becomes much harder and likely infeasible because a malicious transcoder would not know when it should try to submit the fraudulent transcode claim and submit challenged segments for verification.
```
[the remark this answers is not in the document]
Comment (yondonfu) [reply LIPs/LIP-update-solver.md]: Yep, LivepeerVerifier is not proxied so the upgrade process for it just involves deploying a new contract and registering it with the Controller. As a result, none of the old state is migrated, but we also don't have to worry about storage layout when writing the new upgraded contract.

File: LIPs/LIP-update-solver.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+Instead of accepting an array of solver addresses, the constructor will accept a single solver address that will be used as the initially registered solver address. The registered solver address can be updated by `setSolver()` later on by the owner of the Controller.
+
+Let the updated contract be `LivepeerVerifierV2` and the existing deployed contract be `LivepeerVerifierV1`.
+
+A block `upgradeBlock` and round `upgradeRound` (`upgradeBlock` must be a block in `upgradeRound`) are selected. `upgradeRound` is announced, but `upgradeBlock` is not. At a block before `upgradeBlock` within `upgradeRound`, `LivepeerVerifierV2` will be deployed with an address `LivepeerVerifierV2Address`. At block `upgradeBlock`, the owner of the Controller contract will call `Controller.setContract(contractId("Verifier"), LivepeerVerifierV2Address)`. `contractId("Verifier")` returns the keccak256 hash of the provided contract name. Afer the Controller transaction is mined, the upgrade is complete and `LivepeerVerifierV2` will be the new entry point for the centralized transcoding verification system used by the Livepeer protocol.
```
[the remark this answers is not in the document]
Comment (yondonfu) [reply LIPs/LIP-update-solver.md]: `upgradeRound` is just defined for transparency reasons so protocol users can be given a timeframe during which the upgrade described in this proposal will occur. `upgradeBlock` is when the actual upgrade would happen, but in the interest of defending against the mentioned timing attack, it is not announced.

**Comment by yondonfu** [comment]
Re: discussion in an issue prior to creation of PR - I was thinking that discussion in public forums such as Github issues or forum posts would serve as a good way to help develop a specification for a particular proposal when it is at an earlier idea stage. After receiving feedback and exchanging ideas in public forums, the specification for the proposal can be written and the comment thread of the PR would be more focused on the existing content of the written specification as opposed to exploring the general idea space. In this particular case, I already had a specification so I thought it might make sense to skip the public forum discussion and to have the discussion directly in the comment thread of this PR. Let me know what you think!

**Comment by dob** [comment]
Sounds good. One of the meta-challenges I see is that even though you've addressed the issues I raised, and I would now approve the PR, I'm not sure how to determine if there's been enough chance for others to also look and weigh in on the ideas before it becomes accepted. So yes, a forum or issues post about the ideas seem sound in the future to allow that discussion to play out before a PR. But yes, I agree in this case this is more straightforward and was basically ready for spec.

**Comment by yondonfu** [comment]
I think its reasonable to leave this PR open for a few days before merging this as a draft and assigning a number. I'd note that in the current process merging this proposal does not mean that it is is designated "Accepted" status, it just means that it will be a candidate for discussion in a core dev call and if it receives support there it will be designated "Accepted" status. And of course it is up for debate whether this is the right flow for the project at this particular stage, but this is the current outlined approach

**Pull request merged**